include/onewire.h

```diff
@@ -0,0 +1,370 @@
+/**
+ * @file onewire.h
+ * @defgroup onewire onewire
+ * @brief
+ *
+ * Container to store results from telemetry receivied over one wire uart
+ * telemetry
+ */
+
+#pragma once
+#include "dshot.h"
+#include "hardware/uart.h"
+#include "kissesctelem.h"
+#include "stdint.h"
+
+/**
+ * @brief set the number of ESCs at compile time
+ * This is used by the telemetry to tell how many ESCs to read the telemetry
+ * from
+ */
+#ifndef ESC_COUNT
+#define ESC_COUNT 1
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ONEWIRE_BAUDRATE 115200
+
+/**
+ * @brief minimum time interval between telemetry request
+ *
+ * NOTE: minium time interval between telem should be:
+ * min_telem_interval = dshot_interval + telem_delay + telem_data_transfer
+ *                    = 133 us (DSHOT150) + 0 (?)    + 115200 / 10*8 (=695 us)
+ *                    = 827 us
+ * Hence min_telem_interval ~ 1 ms is a safe value
+ */
+static const long int ONEWIRE_MIN_INTERVAL_US = 1000;
+
+/**
+ * @brief Struct to represent ESC data
+ *
+ */
+typedef struct esc_motor {
+  dshot_config *dshot;
+  kissesc_telem_t telem_data;
+} esc_motor_t;
+
+/**
+ * @brief struct to configue one wire uart telemetry
+ *
+ * @ingroup onewire
+ * General config:
+ * @param gpio NOTE: only one gpio because esc telem is one wire
+ * @param send_req_rt Repeating timer config for preodically requesting
+ * telemetry
+ * @param buffer_size = 10 KISS telemetry protocol outputs 10 bytes
+ * Storing uart telemetry output
+ * @param buffer_idx Keep track of idx in uart telemetry buffer
+ * @param buffer Array to hold telemetry buffer data
+ * @param telem_updated_esc -1 <= telem_updated_esc < ESC_COUNT
+ * default -1. If telemetry is received for an ESC, then the ESC idx
+ * is stored in this variable. This idx should be reset
+ * by a main process (e.g. upon reading the onewire data).
+ */
+typedef struct telem_uart {
+  uart_inst_t *uart;
+  uint gpio;
+  uint baudrate;                 /// Defaults to @ref ONEWIRE_BAUDRATE
+  repeating_timer_t send_req_rt; // Repeating timer config for periodically
+                                 // requesting telemetry
+  bool send_req_rt_state;        // Keep track of repeating timer state
+  volatile size_t esc_motor_idx; // Keep track of which esc to request telem
+  volatile esc_motor_t escs[ESC_COUNT];
+
+  // variable to store the result from a telemetry read in an irq
+  volatile size_t buffer_idx;
+  volatile uint8_t buffer[KISS_ESC_TELEM_BUFFER_SIZE];
+  // flag is updated when esc telemetry has been received
+  int telem_updated_esc;
+} onewire_t;
+
+// Global variable for onewire
+onewire_t onewire;
```

Comment:
```suggestion
extern onewire_t onewire;
```

You should not define global variables in header files. You should define them in .c source file.

If global variable is to be visible within only one .c file, you should declare it static.
If global variable is to be used across multiple .c files, you should not declare it static. Instead you should declare it extern in header file included by all .c files that need it.

Reply:
Thanks for taking a look. Just making that change now and adding the variable to a source file. It's interesting that `static` worked for you earlier when we tested it, because theoretically static variables shouldn't be visible outside that file.

Reply:
Fixed in https://github.com/Guppy16/pico-dshot/pull/3/commits/5597b5606dc6556372edbba1eac312295ac0f3ae